### PR TITLE
[USMON-668] Moves connection tuple normalization to https entrypoint rather than the TLS hooks

### PR DESCRIPTION
**Pull request by Yumasi**
### What does this PR do?

The PR moves the connection tuple normalization out of the TLS hooks and moves it to the HTTPs entrypoint.

### Motivation

Preliminary change required for HTTP2-TLS support, as we need to have unnormalized connection, while HTTPs does require normalization

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/tls/https.h [hunk 1]
```diff
@@ -76,6 +76,7 @@ static __always_inline void tls_process(struct pt_regs *ctx, conn_tuple_t *t, vo
     switch (protocol) {
     case PROTOCOL_HTTP:
         prog = TLS_HTTP_PROCESS;
+        normalize_tuple(t);
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/tls/https.h:79]: please add a documentation explaining the need for that

File: pkg/network/ebpf/c/runtime/usm.c [hunk 2]
```diff
@@ -262,10 +262,13 @@ int uprobe__crypto_tls_Conn_Read__return(struct pt_regs *ctx) {
...
-    char *buffer_ptr = (char*)call_data_ptr->b_data;
-    bpf_map_delete_elem(&go_tls_read_args, (go_tls_function_args_key_t*)&call_key);
+    conn_tuple_t copy = *t;
+    flip_tuple(&copy);
+
+    char *buffer_ptr = (char *)call_data_ptr->b_data;
+    bpf_map_delete_elem(&go_tls_read_args, (go_tls_function_args_key_t *)&call_key);
 
-    tls_process(ctx, t, buffer_ptr, bytes_read, GO);
+    tls_process(ctx, &copy, buffer_ptr, bytes_read, GO);
```
Comment (guyarb) [inline pkg/network/ebpf/c/runtime/usm.c:271]: please document, it is unclear for the naive reader

Comment (guyarb) [reply pkg/network/ebpf/c/runtime/usm.c:271]: also, try to limit the change, by moving the creation of `copy` to be just on top of `tls_process`

**Review by guyarb — CHANGES_REQUESTED** [review]

**Comment by guyarb** [comment]
1. please elaborate the motivation for the change

2. also I think the line below is inaccurate
> This PR adds a conn tuple flip in the GoTLS read probe, so that it can be identical to the tuple from the write probe

3. What about the scenario of both client and server running locally (a.k.a collision when the PID is being zeroed)

**Review by guyarb — DISMISSED** [review]
Reviewed again the code - the same problem will happen with openssl as well as in `tup_from_ssl_ctx` (the equivalent function of `conn_tup_from_tls_conn`) does the same flip

Also, we're not directly normalizing the tuple, but 
```c
    if (!is_ephemeral_port(conn_tuple.sport)) {
        flip_tuple(&conn_tuple);
    }
```
while normalize_tuple is
```c
static __always_inline bool normalize_tuple(conn_tuple_t *t) {
    if (is_ephemeral_port(t->sport) && !is_ephemeral_port(t->dport)) {
        return false;
    }

    if ((!is_ephemeral_port(t->sport) && is_ephemeral_port(t->dport)) || t->dport > t->sport) {
        // flip the tuple if:
        // 1) the tuple is currently in the (server, client) format;
        // 2) unlikely: if both ports are in the same range we ensure that sport > dport to make
        // this function return a deterministic result for a given pair of ports;
        flip_tuple(t);
        return true;
    }

    return false;
}
```

File: pkg/network/ebpf/c/runtime/usm.c [hunk 3]
```diff
@@ -262,10 +262,17 @@ int uprobe__crypto_tls_Conn_Read__return(struct pt_regs *ctx) {
...
-    bpf_map_delete_elem(&go_tls_read_args, (go_tls_function_args_key_t*)&call_key);
+    char *buffer_ptr = (char *)call_data_ptr->b_data;
+    bpf_map_delete_elem(&go_tls_read_args, (go_tls_function_args_key_t *)&call_key);
+
+    // We need to flip the tuple to match the tuple from the write uprobe. This
+    // is needed for HTTP2 decoding to work properly, as without that tuple flip,
+    // we would use different dynamic tables for decoding the buffers from the read
+    // & write uprobes.
+    conn_tuple_t copy = *t;
+    flip_tuple(&copy);
```
Comment (DanielLavie) [inline pkg/network/ebpf/c/runtime/usm.c:273]: Isn't it a TLS hook, hence we want to avoid the tuple normalization?

**Review by DanielLavie — APPROVED** [review]

File: pkg/network/ebpf/c/runtime/usm.c (hunk 3 above)
Comment (guyarb) [reply pkg/network/ebpf/c/runtime/usm.c:273]: sorry it should have been removed

**Pull request merged**